File: fundamentals/Drill.java
// ============================================================================
// Drill.java - fill in each method from memory, then:
//     javac Drill.java Start.java && java Start        (or:  ./run.sh)
//
// No reference until you've been stuck >90 seconds. On a peek, retype the WHOLE
// method. The timer starts on your first `java Start` and persists across
// re-runs in a hidden .drill.start marker, resetting on a full CORE clear.
// Start.java (the runner) is separate - don't edit it.
// ============================================================================
import java.util.*;

public class Drill {

    public static class ListNode {
        int val; ListNode next;
        ListNode() {}
        ListNode(int v) { val = v; }
        ListNode(int v, ListNode n) { val = v; next = n; }
    }
    public static class TreeNode {
        int val; TreeNode left, right;
        TreeNode() {}
        TreeNode(int v) { val = v; }
        TreeNode(int v, TreeNode l, TreeNode r) { val = v; left = l; right = r; }
    }

    // 1. Reverse a linked list. Return the new head.
    public static ListNode reverseList(ListNode head) {
        return null; // your code
    }

    // 2. Delete the middle node (floor(n/2)-th, 0-indexed). [1,2,3,4]->[1,2,4]. [1]->[].
    public static ListNode deleteMiddleNode(ListNode head) {
        return null; // your code
    }

    // 3. Odd/Even list: regroup odd POSITIONS first, then even. [1,2,3,4,5]->[1,3,5,2,4].
    public static ListNode oddEvenList(ListNode head) {
        return null; // your code
    }

    // 4. Iterative DFS with an explicit stack. Visit order from start.
    public static List<Integer> dfsIterative(Map<Integer, List<Integer>> graph, int start) {
        return new ArrayList<>(); // your code
    }

    // 5. BFS over an adjacency-list graph. First-visit order from start.
    public static List<Integer> bfs(Map<Integer, List<Integer>> graph, int start) {
        return new ArrayList<>(); // your code
    }

    // 6. Recursive DFS. Visit order from start.
    public static List<Integer> dfsRecursive(Map<Integer, List<Integer>> graph, int start) {
        return new ArrayList<>(); // your code
    }

    // 7. Top K largest, sorted descending. Use a heap.
    public static List<Integer> topK(int[] nums, int k) {
        return new ArrayList<>(); // your code
    }

    // 8. Binary search. Index of target in a sorted array, else -1.
    public static int binarySearch(int[] nums, int target) {
        return -1; // your code
    }

    // 9. Leftmost insertion point: index of the first element >= target.
    public static int lowerBound(int[] nums, int target) {
        return 0; // your code
    }

    // 10. Fixed-size sliding window. Max sum of any window of size k, in one pass.
    public static int maxWindowSum(int[] nums, int k) {
        return 0; // your code
    }

    // 11. Variable sliding window. Longest substring with no repeats.
    public static int longestUniqueSubstring(String s) {
        return 0; // your code
    }

    // 12. Max depth of a binary tree. Empty -> 0.
    public static int maxDepth(TreeNode root) {
        return 0; // your code
    }

    // 13. Level-order traversal -> list of levels.
    public static List<List<Integer>> levelOrder(TreeNode root) {
        return new ArrayList<>(); // your code
    }

    // 14. Validate BST (strict, bounded by ancestors).
    public static boolean isValidBst(TreeNode root) {
        return false; // your code
    }

    // 15. Search in a BST. Return the subtree whose value == val, else null.
    public static TreeNode searchBst(TreeNode root, int val) {
        return null; // your code
    }

    // 16. Group anagrams. Key each word by its sorted letters.
    public static List<List<String>> groupAnagrams(String[] strs) {
        return new ArrayList<>(); // your code
    }

    // 17. Number of islands (4-directional) in a grid of 0s and 1s.
    public static int numIslands(int[][] grid) {
        return 0; // your code
    }

    // 18. 3Sum. All unique triplets summing to 0.
    public static List<List<Integer>> threeSum(int[] nums) {
        return new ArrayList<>(); // your code
    }

    // 19. Daily temperatures. Days until a warmer temp (0 if none).
    public static int[] dailyTemperatures(int[] temperatures) {
        return new int[0]; // your code
    }

    // 20. Merge overlapping [start,end] intervals (unsorted input).
    public static int[][] mergeIntervals(int[][] intervals) {
        return new int[0][]; // your code
    }

    // 21. Delete a node with value key in a BST; return the (possibly new) root.
    public static TreeNode deleteBstNode(TreeNode root, int key) {
        return null; // your code
    }

    // 22. Subsets / power set of distinct nums.
    public static List<List<Integer>> subsets(int[] nums) {
        return new ArrayList<>(); // your code
    }

    // 23. Koko / min eating speed. Smallest integer speed to eat all piles within h.
    public static int minEatingSpeed(int[] piles, int h) {
        return 0; // your code
    }

    // 24. Subarray sum equals k (count). Seed {0:1}.
    public static int subarraySum(int[] nums, int k) {
        return 0; // your code
    }

    // ████████████████████████████████████████████████████████████████████████
    // BONUS - extra reps; not scored, never gate the timer.
    // ████████████████████████████████████████████████████████████████████████

    // B1. Coin change (fewest coins), or -1.
    public static int coinChange(int[] coins, int amount) {
        return -1; // your code
    }

    // B2. Course schedule: true if all courses can finish. [a,b] means b before a.
    public static boolean courseSchedule(int numCourses, int[][] prerequisites) {
        return false; // your code
    }

    // B3. Unique right/down paths in an m x n grid.
    public static int uniquePaths(int m, int n) {
        return 0; // your code
    }

    // B4. Diameter of a binary tree (edges on longest path).
    public static int diameterOfBinaryTree(TreeNode root) {
        return 0; // your code
    }
}

File: fundamentals/drill.cpp
// ============================================================================
// drill.cpp - fill in each function from memory, then:
//     g++ -std=c++17 drill.cpp -o drill && ./drill      (or:  ./run.sh)
//
// No reference until you've been stuck >90 seconds. On a peek, retype the WHOLE
// function. The timer starts on your first ./drill and persists across re-runs.
// harness.hpp (the runner) is included at the bottom - don't edit it.
// ============================================================================
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val; ListNode* next;
    ListNode(int v = 0, ListNode* n = nullptr) : val(v), next(n) {}
};
struct TreeNode {
    int val; TreeNode* left; TreeNode* right;
    TreeNode(int v = 0, TreeNode* l = nullptr, TreeNode* r = nullptr) : val(v), left(l), right(r) {}
};

// ============================================================================
// 1. Reverse a linked list. Return the new head.
// ============================================================================
ListNode* reverse_list(ListNode* head) {
    return nullptr; // your code
}

// ============================================================================
// 2. Delete the middle node (floor(n/2)-th, 0-indexed). [1,2,3,4]->[1,2,4]. [1]->[].
// ============================================================================
ListNode* delete_middle_node(ListNode* head) {
    return nullptr; // your code
}

// ============================================================================
// 3. Odd/Even list: regroup odd POSITIONS first, then even. [1,2,3,4,5]->[1,3,5,2,4].
// ============================================================================
ListNode* odd_even_list(ListNode* head) {
    return nullptr; // your code
}

// ============================================================================
// 4. Iterative DFS with an explicit stack. Visit order from start.
// ============================================================================
vector<int> dfs_iterative(unordered_map<int, vector<int>>& graph, int start) {
    return {}; // your code
}

// ============================================================================
// 5. BFS over an adjacency-list graph. First-visit order from start.
// ============================================================================
vector<int> bfs(unordered_map<int, vector<int>>& graph, int start) {
    return {}; // your code
}

// ============================================================================
// 6. Recursive DFS. Visit order from start.
// ============================================================================
vector<int> dfs_recursive(unordered_map<int, vector<int>>& graph, int start) {
    return {}; // your code
}

// ============================================================================
// 7. Top K largest, sorted descending. Use a heap.
// ============================================================================
vector<int> top_k(vector<int> nums, int k) {
    return {}; // your code
}

// ============================================================================
// 8. Binary search. Index of target in a sorted array, else -1.
// ============================================================================
int binary_search(vector<int>& nums, int target) {
    return -1; // your code
}

// ============================================================================
// 9. Leftmost insertion point: index of the first element >= target.
// ============================================================================
int lower_bound_idx(vector<int>& nums, int target) {
    return 0; // your code
}

// ============================================================================
// 10. Fixed-size sliding window. Max sum of any window of size k, in one pass.
// ============================================================================
int max_window_sum(vector<int>& nums, int k) {
    return 0; // your code
}

// ============================================================================
// 11. Variable sliding window. Longest substring with no repeats.
// ============================================================================
int longest_unique_substring(string s) {
    return 0; // your code
}

// ============================================================================
// 12. Max depth of a binary tree. Empty -> 0.
// ============================================================================
int max_depth(TreeNode* root) {
    return 0; // your code
}

// ============================================================================
// 13. Level-order traversal -> list of levels.
// ============================================================================
vector<vector<int>> level_order(TreeNode* root) {
    return {}; // your code
}

// ============================================================================
// 14. Validate BST (strict, bounded by ancestors).
// ============================================================================
bool is_valid_bst(TreeNode* root) {
    return false; // your code
}

// ============================================================================
// 15. Search in a BST. Return the subtree whose value == val, else nullptr.
// ============================================================================
TreeNode* search_bst(TreeNode* root, int val) {
    return nullptr; // your code
}

// ============================================================================
// 16. Group anagrams. Key each word by its sorted letters.
// ============================================================================
vector<vector<string>> group_anagrams(vector<string>& strs) {
    return {}; // your code
}

// ============================================================================
// 17. Number of islands (4-directional) in a grid of 0s and 1s.
// ============================================================================
int num_islands(vector<vector<int>> grid) {
    return 0; // your code
}

// ============================================================================
// 18. 3Sum. All unique triplets summing to 0.
// ============================================================================
vector<vector<int>> three_sum(vector<int> nums) {
    return {}; // your code
}

// ============================================================================
// 19. Daily temperatures. Days until a warmer temp (0 if none).
// ============================================================================
vector<int> daily_temperatures(vector<int>& temperatures) {
    return {}; // your code
}

// ============================================================================
// 20. Merge overlapping [start,end] intervals (unsorted input).
// ============================================================================
vector<vector<int>> merge_intervals(vector<vector<int>> intervals) {
    return {}; // your code
}

// ============================================================================
// 21. Delete a node with value key in a BST; return the (possibly new) root.
// ============================================================================
TreeNode* delete_bst_node(TreeNode* root, int key) {
    return nullptr; // your code
}

// ============================================================================
// 22. Subsets / power set of distinct nums.
// ============================================================================
vector<vector<int>> subsets(vector<int>& nums) {
    return {}; // your code
}

// ============================================================================
// 23. Koko / min eating speed. Smallest integer speed to eat all piles within h.
// ============================================================================
int min_eating_speed(vector<int>& piles, int h) {
    return 0; // your code
}

// ============================================================================
// 24. Subarray sum equals k (count). Seed {0:1}.
// ============================================================================
int subarray_sum(vector<int>& nums, int k) {
    return 0; // your code
}

// ████████████████████████████████████████████████████████████████████████████
// BONUS - extra reps; not scored, never gate the timer.
// ████████████████████████████████████████████████████████████████████████████

// B1. Coin change (fewest coins), or -1.
int coin_change(vector<int>& coins, int amount) {
    return -1; // your code
}

// B2. Course schedule: true if all courses can finish. [a,b] means b before a.
bool course_schedule(int num_courses, vector<vector<int>>& prerequisites) {
    return false; // your code
}

// B3. Unique right/down paths in an m x n grid.
int unique_paths(int m, int n) {
    return 0; // your code
}

// B4. Diameter of a binary tree (edges on longest path).
int diameter_of_binary_tree(TreeNode* root) {
    return 0; // your code
}

// ---- runner (do not edit) --------------------------------------------------
#include "harness.hpp"
int main() { return run_drills(); }
